analyzers/src/SonarAnalyzer.Common/Rules/CommentsShouldNotBeEmptyBase.cs

```diff
@@ -0,0 +1,145 @@
+                CheckTrivia(c, token.LeadingTrivia);
+                CheckTrivia(c, token.TrailingTrivia);
+            }
+        });
+
+    private void CheckTrivia(SonarSyntaxTreeReportingContext context, IEnumerable<SyntaxTrivia> trivia)
+    {
+        var partitions = Partition(trivia);
+        if (partitions is null)
+        {
+            return;
+        }
+
+        foreach (var partition in partitions.Where(ShouldReport))
+        {
+            var start = partition.First().GetLocation().SourceSpan.Start;
+            var end = partition.Last().GetLocation().SourceSpan.End;
+            var location = Location.Create(context.Tree, TextSpan.FromBounds(start, end));
+            context.ReportIssue(Diagnostic.Create(Rule, location));
+        }
+
+        bool ShouldReport(IEnumerable<SyntaxTrivia> trivia) =>
+            trivia.Any() && trivia.All(x => string.IsNullOrWhiteSpace(GetCommentText(x)));
+    }
+
+    private List<List<SyntaxTrivia>> Partition(IEnumerable<SyntaxTrivia> trivia)
+    {
+        // hot path: avoid unnecessary allocations
+        List<List<SyntaxTrivia>> res = null;
+        List<SyntaxTrivia> current = null;
+        var firstEndOfLineFound = false;
+
+        foreach (var trivium in trivia)
+        {
+            if (IsSimpleComment(trivium)) // put it on the current block of "//"
+            {
+                current ??= new();
+                current.Add(trivium);
+                firstEndOfLineFound = false;
+            }
+            // This is for the case, of two different comment types, for example:
+            // //
+            // ///
+            else if (IsValidTriviaType(trivium)) // valid but not "//", because of the upper if
+            {
+                CloseCurrentPartition();
+                // all comments except single-line comments are parsed as a block already.
+                current ??= new();
+                current.Add(trivium);
+                CloseCurrentPartition();
+            }
+            // This handles an empty line, for example:
+            // // some comment \n <- EOL found, set to true
+            // //  \n <- EOL is set to false at CommentTrivia, set to true after it
+            // // some other comment <- EOL is set to false at CommentTrivia, set to true after it
+            // \n <- EOL found, is already true, closes current partition
+            else if (IsEndOfLine(trivium))
+            {
+                if (firstEndOfLineFound)
+                {
+                    CloseCurrentPartition();
+                }
+                else
+                {
+                    firstEndOfLineFound = true;
+                }
+            }
+            else if (!IsWhitespace(trivium))
+            {
+                CloseCurrentPartition();
+            }
+        }
+
+        if (current is not null)
+        {
+            res ??= new();
+            res.Add(current);
+        }
+
+        return res;
+
+        void CloseCurrentPartition()
```

Comment:
This captures current and res. This might result in a capture class being created in the background and allocated for each call to PartitionComments. You can avoid this by doing
```cs
// Hotpath: Don't capture variables
static void CloseCurrentPartition(ref List<Trivia> current, ref List<List<Trivia>> res)
```